### DISCUSSION TITLE
PacketQueue implementation in Ieee802154Mac module

### DISCUSSION THREAD

**Discussion by n-jay** [omnetpp/omnetpp#1124] [Q&A] [answered]
Hello everyone,
I am currently working on implementing memory buffers in nodes that use the `Ieee802154NarrowbandInterface`, which utilizes the `Ieee802154Mac` module. I am using **INET 4.2.5**, so the naming conventions might differ from the latest version.

I have been experimenting with the `PacketQueue` simple module that is implemented within `Ieee802154Mac`, and I need clarification on the following:

1. whether the `PacketQueue` module acts as a buffer for _transmissions_, _receptions_, or _both_.  Just observing the packet flow using the Qtenv GUI and it seems like both. 
2. If _only transmissions_, do we have to add _another queue_ as a submodule to collect receiving signals?
3. When the **packetPushed** statistic is checked it is significantly less than the number of packets that have been registered by a module in the network work. Does that mean not all packets received at a node is getting registered in the queue?

**Comment by n-jay** [comment]
@rhornig would appreciate some thoughts on this. 😄

**Reply by rhornig** [reply]
Sadly, this is above may pay grade :) @levy perhaps?

**Comment by levy** [comment, marked as answer]
The PacketQueue module is just a queue for packets, so in itself it's neither for transmitted or received packets. In network interface modules, the queue submodule, that is a PacketQueue module by default, is used to store outgoing packets. In other words the queue module stores packets that will be transmitted. Please note that there are as many PacketQueue modules as there are network interfaces in the network. This may be the answer for your third question.

**Reply by n-jay** [reply]
Noted @levy, thanks!
In relation to Q2, do we have to add another queue as a submodule (within `Ieee802154Mac` module) to store receiving packets? or can the existing queue be modified to accept receiving packets also?
